tabbycat/venues/allocator.py

```diff
@@ -168,7 +196,139 @@ def allocate_unconstrained_venues(self, debates):
         random.shuffle(debates)
         return {debate: venue for debate, venue in zip(debates, self._preferred_venues)}
 
-    def save_venues(self, debate_venues):
-        for debate, venue in debate_venues.items():
-            debate.venue = venue
-        Debate.objects.bulk_update(debate_venues.keys(), ['venue'])
+
+class BaseHungarianVenueAllocator(BaseVenueAllocator):
+    """Base class for venue allocations using the Hungarian algorithm.
+
+    Most of the costs would remain the same between allocators. Works as an
+    allocator on its own (no cost for history)"""
+
+    description = _("Hungarian")
+
+    def __init__(self, round, debates=None):
+        super().__init__(round, debates)
+
+        t = self.round.tournament
+        self.history_cost = t.pref('venue_history_cost')
+        self.constraint_cost = t.pref('venue_constraint_cost')
+        self.bad_venue_cost = t.pref('venue_score_cost')
+
+        self.munkres = Munkres()
+
+    def get_debate_queryset(self):
+        """For history constraints.
+
+        Get previous debates (with venues & categories) of participants in addition to current ones"""
+
+        return self.round.debate_set.all().select_related('venue').prefetch_related(
+            Prefetch('debateadjudicator_set', queryset=DebateAdjudicator.objects.select_related('adjudicator', 'adjudicator__institution')),
+            Prefetch('debateteam_set', queryset=DebateTeam.objects.select_related('team', 'team__institution')),
+        )
+
+    def allocate(self):
+        self.prev_venues = self.get_previous_venue_categories()
+        self.constraints = self.collect_dict_constraints()
+
+        self.max_venue_score = self.venues[0].priority
+        self.max_constraint_priority = list(self.constraints.values())[0][0].priority
+
+        cost_matrix = []
+        for d in self.debates:
+            cost_matrix.append(self.debate_cost_calc(d))
+
+        indices = self.munkres.compute(cost_matrix)
+
+        result = {self.debates[i]: self.venues[j] for i, j in indices}
+        self.save_venues(result)
+
+    def get_previous_venue_categories(self):
+        debate_prefetch = Prefetch('debateteam_set', queryset=DebateTeam.objects.filter(
+            debate__round__seq__lt=self.round.seq, debate__venue__isnull=False, debate__venue__venuecategory__rotate=True).distinct())
+        teams = self.round.active_teams.prefetch_related(debate_prefetch)
+
+        prev_team_vc = {}
+        for t in teams.all():
+            prev_team_vc[t] = []
+            for dt in t.debateteam_set.all():
+                prev_team_vc[t].extend(dt.debate.venue.venuecategory_set.all())
+
+        self.prev_venue_categories = {}
+        for debate in self.debates:
+            self.prev_venue_categories[debate] = []
+
+            # Ajudicators are excepted from this constraint
+            for dt in debate.debateteam_set.all():
+                self.prev_venue_categories[debate].extend(prev_team_vc[dt.team])
```

Comment:
Is "previous value" functionality unused in the non-rotation Hungarian allocators? It might make the code a bit easier to understand if this is spun out to an intermediate base class, so like:

```
                                                                RotationVenueAllocator
                                                              /
HungarianVenueAllocator  <-- BaseHistoryHungarianVenueAllocator
                                                              \
                                                                StationaryVenueAllocator
```

(I realise this makes the class structure more complicated, but it did confuse me for a bit.)

Also, I try to name classes `Base` if and only if they're abstract base classes (not to be used directly); if it can be used directly, it's fine to not call it `Base` even if lots of things derive from it.